README.md
# ModelDDvlPlone
Python/Zope/Plone product: Model Driven runtime support framework by MDDsl that animates the Plone products gvSIG-i18n, gvSIG-business process definition (gvSIG-bpd), ModelDDploneMOF, ModelDDploneRDB, ModelDDploneBMM. Runs on Zope2.9/Plone2.5. 

Overview

  ModelDDvlPlone and its companion ModelDDvlPloneTool
  provide respectively the skins (visual presentation)
  and logic (retrieve, modify) to manage the creation of 
  complex Archetypes networks, including their linkage
  through relations.
  
  
Dependencies

  ModelDDvlPlone and its companion ModelDDvlPloneTool
  require the installation of the products:
  
  - "Archetypes"
  
  - "Relations"
  
  - "PloneLanguageTool"
  
  Installation will fail if these products are not 
  correctly deployed and available in the Plone instance.
  
  The installer will automatically install the prerequisite
  products, if corretly deployed and available in the Plone instance.
  
  
Copyright

  Copyright (c) 2008 by 2008 Model Driven Development sl and Antonio Carrasco Valero
  

Project Home

  - "ModelDD.org":http://www.ModelDD.org 

  
  
Authors

  - "Antonio Carrasco Valero and "Model Driven Development sl  Valencia (Spain)":http://www.ModelDD.org 

  
License

  GNU General Public License (GPL)
 
  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.
 
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.
 
  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
  02110-1301, USA.


Relations/ChangeLog
2006-03-30  Daniel Nouri  <daniel (dot) nouri at gmail (dot) com>

	* utils.py (adddeleteVocab): I put the ``adddelete_vocab`` script
	onto the filesystem while adding a check for
	``ModifyPortalContent`` on context.  The Python script still
	exists, but it's only an alias now.  Also added tests, which the
	script didn't have.

2006-03-01  Jens Klein <jens@bluedynamics.com>

        * field.py: i18n moved in Archetypes 1.4 to 
	Products.Archetypes.generator.i18n

2006-02-04  Daniel Nouri  <dpunktnpunkt at web.de>

	* tests: Added more tests to check if multiple refs per triple
	(i.e. same source, target, relationship) are working.  (Note that
	this currently requires the dpunktnpunkt-multipleref branch of
	Archetypes and that it's disallowed by default.

2006-01-28  Daniel Nouri  <dpunktnpunkt at web.de>

	* components/inverse.py (InverseImplicator.implyOnDisconnect):
	Improvement of implementation.

2006-01-26  Daniel Nouri  <dpunktnpunkt at web.de>

	* components/inverse.py (InverseImplicator): New implementation
	that no longer assumes that there's only one reference with the
	same relationship between two objects.

	* interfaces.py (IRuleset.implyOnConnect): No longer care about
	whether a reference with the same triple of (source, target,
	relationship) already exists.

	* config.py (ALLOW_MULTIPLE_REFS_PER_TRIPLE): Added configuration
	option for updateReferences argument of
	ReferenceEngine.addReference.  It's turned off by default.

	* interfaces.py (IRuleset.implyOnConnect)
	(IPrimaryImplicator.connect): Added metadata keyword argument that
	lets you provide attributes for the reference.  Note that this is
	subject to change and will probably be moved out of the public
	API.

	* interfaces.py (IReferenceConnectionProcessor.process): Elements
	of ``disconnect`` may also be UIDs of reference objects.

2005-11-22 Jens Klein <jens@bluedynamics.com>

	* fixed hiding of tools from navigation and tabs

2005-08-10  Daniel Nouri  <dpunktnpunkt at web.de>

	* ruleset.py (Library.__implements__): Fixed the faulty
	IActionProvider implementation of Library.

2005-07-20  Daniel Nouri  <dpunktnpunkt at web.de>

	* components/cardinality.py: Applied patch by David Baehrens that
	allows us to define source cardinality with CardinalityConstraint.

2005-07-07  Daniel Nouri  <dpunktnpunkt at web.de>

	* ruleset.py (XMLImportExport.schema): Use the accessor as the
	default_method.

	* utils.py (AllowedTypesByIface._verifyObjectPaste): Use
	PortalFolderBase's _verifyObjectPaste as we're actually not
	inheriting from PortalFolder.

	* tests/testRuleset.py (TestRuleset.testRenameRulesetInLibrary):
	Import transaction instead of using the global get_transaction.

	* processor.py (process): Import transaction instead of using the
	global get_transaction.
	
2005-06-03  Daniel Nouri  <dpunktnpunkt at web.de>

	* components/contentreference.py (ContentReferenceFinalizer):
	Added isPrimary attribute to CRF.

2005-05-28  Daniel Nouri  <dpunktnpunkt at web.de>

	* Extensions/Install.py (install_tools): Don't add library to
	metaTypesNotToList.

2005-05-18  Daniel Nouri  <dpunktnpunkt at web.de>

	* skins/relations/relations_adddelete_vocab.py, model/*, tests/relations_sample.xml, tests/testXMLImportExport.py, doc/Overview.txt, interfaces.py, ruleset.py, field.py:
	Merged relations_xml_import_export branch.  Adds import/export
	capabilities to Rules, Rulesets, Library and RulesetCollections.
	
2005-02-21  Daniel Nouri  <dpunktnpunkt at web.de>

	* interfaces.py (IRuleset.getComponents): This replaces the earlier
	private _filterByInterface.

	* interfaces.py (IRuleset.listActionsFor): Added a method which I
	forgot to put in the interface earlier.

	* interfaces.py (IVocabularyProvider.getSearchTerms): This should allow
	us to interface with ATReferenceBrowserWidget in a reasonable manner.

	* components/types.py (PortalTypeConstraint.getSearchTerms):
	Comply with IVocabularyProvider.getSearchTerms .
	
2005-02-16  Daniel Nouri  <dpunktnpunkt at web.de>

	* doc/Overview.txt: Doctest examples.

2005-02-06  Daniel Nouri  <dpunktnpunkt at web.de>

	* doc/Overview.txt: Added Developer's Documentation.  More to
	come.
	
	* interfaces.py: Cleaned up a bit.
	
2005-02-02  Daniel Nouri  <dpunktnpunkt at web.de>

	* field.py: Slimmed field.py by removing reimplemented methods
	_Vocabulary and Vocabulary of ReferenceField.  The field's purpose
	needs to be made clear, still.  See XXX comments.
	
2005-01-27  Daniel Nouri  <dpunktnpunkt at web.de>

	* ruleset.py (RulesetCollection): Implemented a "Ruleset Collection"

Relations/ruleset.py
# ###################################################################################
# ACV OJO FIX 2008/04/30
#     fails to load multiple allowedSourceType elements or multiple allowedTargetType elements
# SEE BELOW in ruleset.XMLImportDOM.importDOM() 
#

from xml.dom import minidom

from Acquisition import ImplicitAcquisitionWrapper, aq_parent, aq_inner
from Globals import InitializeClass
import zExceptions
import transaction

from OFS.ObjectManager import BeforeDeleteException
from Products import CMFCore
from Products.CMFCore.utils import getToolByName
from Products.CMFCore.ActionProviderBase import ActionProviderBase
from Products.CMFCore.Expression import Expression
from Products.Archetypes.interfaces.referenceable import IReferenceable
from Products.Archetypes.Referenceable import Referenceable
from Products.Archetypes.ReferenceEngine import Reference
from Products.Archetypes.utils import shasattr, getRelURL
from Products.Archetypes.public import *

from config import *
from Products.Relations import config
import interfaces
import brain
import schema
import utils

class XMLImportExport:
    __implements__ = interfaces.IXMLImportExport,

    schema=Schema((
        StringField('xml',
                    mutator='importXML',
                    accessor='exportXML',
                    default_method='exportXML',
                    widget=TextAreaWidget(rows=20),
                    schemata='xml',
                    mode='w',
                    ))
                  )

    def importXML(self, xmlstring):
        doc=minidom.parseString(xmlstring)
        #extract the main element node
        element=[el for el in doc.childNodes
                 if el.nodeType==doc.ELEMENT_NODE][0]
        return self.importDOM(doc, element)
    
    def importDOM(self, doc, element, root=1):
        if root:
            doc.objectsByUid={}
            doc.references=[]
            
        childnodes = [node for node in element.childNodes
                      if node.nodeType == element.ELEMENT_NODE]
            
        title=element.getAttribute('title') or self.getId()
        self.setTitle(title)
        
        # counter to enable subtransactions on ever X imported rulesets
        counter = 0
            
        for node in childnodes:
            counter += 1
            tagname=node.tagName
            if tagname in self.Schema().keys():
                key=tagname
                ref=self.Schema()[key]
                
                if ref.type!='reference':
                    value=node.firstChild.nodeValue.strip()
                    if value =='None':
                        value=None
                        
                    self.update(**{str(key):value})
                else:
                    doc.references.append(
                        (self, ref, node.getAttribute('uidref').strip()))
                    
            elif tagname+'s' in self.Schema().keys():
                key=tagname+'s'

# ###################################################################################
# ACV OJO FIX 2008/04/30
#
#   Problem:
#       ruleset.XMLImportDOM.importDOM() fails to load multiple allowedSourceType elements or multiple allowedTargetType elements
#
#   Prerequisites:
#       Installed Module: Relations 0.6b 
#
#   When:
#       In a Plone Site, which as already installed Relations 0.6b
#       Installing a new Product which utilizes the Relations functionality
#       by including a relations.xml file
#       with a RuleSetCollection element with a RuleSet element with a TypeConstraint element
#       with multiple allowedSourceType elements or multiple allowedTargetType elements
#
#   Error log:
#       Module Products.CMFQuickInstallerTool.QuickInstallerTool, line 322, in installProduct
#           - __traceback_info__: ('gvSigPPP',)
#       Module Products.ExternalMethod.ExternalMethod, line 225, in __call__
#           - __traceback_info__: ((<PloneSite at /p1>,), {}, (False,))
#       Module D:\dvpt\Plone251\Data\Products\gvSigPPP\Extensions\Install.py, line 88, in install
#       Module Products.Relations.ruleset, line 45, in importXML
#       Module Products.Relations.ruleset, line 110, in importDOM
#       Module Products.Relations.ruleset, line 110, in importDOM
#       Module Products.Relations.ruleset, line 110, in importDOM
#       Module Products.Relations.ruleset, line 80, in importDOM
#       TypeError: can only concatenate tuple (not "list") to tuple
#
#   Sample input:
#		<Ruleset id="Guias_detallan_Reglamento" uid="-64--88-1-33-76a6eb09:1199c527f27:-8000:0000000000000DD0">
#			<TypeConstraint id="type_constraint">
#				<allowedSourceType>
#					Guia
#				</allowedSourceType>
#				<allowedTargetType>
#					Reglamento
#				</allowedTargetType>
#				<allowedTargetType>
#					Procedimento
#				</allowedTargetType>
#				<allowedTargetType>
#					ProcedimentoSimple
#				</allowedTargetType>
#				<allowedTargetType>
#					Norma
#				</allowedTargetType>
#			</TypeConstraint>
#			<InverseImplicator id="inverse_relation">
#				<inverseRuleset uidref="reglamentos_guias"/>
#			</InverseImplicator>
#		</Ruleset>
#
#   Diagnosis:
#       The first allowedSourceType or allowedTargetType is registered in self dictionary as a tuple, not as a list
#       To register additional allowedSourceType or allowedTargetType the concatenation of the new element list
#       with the existing one fails with error TypeError: can only concatenate tuple (not "list") to tuple
#
#   Fix:
#      To convert the node value tuple already existing in the self dictionary, when not empty,
#      to a list that can be concatenated with the just read new element list
#
#   Previous code:
#                self.update(
#                    ** { str(key):
#                         (self.Schema().get(key).get(self) or []) +
#                         [node.firstChild.nodeValue]
#                       }
#                    )                
#
#                log("\n\n!!!ACVOJO!!!  ruleset.importDOM()" )
#                log("!!!ACVOJO!!! key %s >>>" % str( key) )
#                log("!!!ACVOJO!!! self.Schema().get(key).get(self) %s >>>" % str( self.Schema().get(key).get(self)) )
#                log("!!!ACVOJO!!! self.Schema().get(key) %s >>>" % str( self.Schema().get(key)) )
#                log("!!!ACVOJO!!! self.Schema() %s >>>" % str( self.Schema()) )
#                log("!!!ACVOJO!!! [node.firstChild.nodeValue] %s >>>" % str( [node.firstChild.nodeValue]) )
#                log("!!!ACVOJO!!! [node.firstChild] %s >>>" % str( [node.firstChild]) )
#                log("!!!ACVOJO!!! [node] %s >>>" % str( [node]) )
#
                
                if self.Schema().get(key).get(self):
                   unExistingNodeValue=[aV for aV in self.Schema().get(key).get(self)]
                else:
                   unExistingNodeValue=[]

#
#                log("!!!ACVOJO!!! unACVOJOValue %s >>>" % str( unACVOJOValue) )
#
               

                self.update(
                    ** { str(key):
# ACV OJO FIX replaced line         (self.Schema().get(key).get(self) or []) +
                         unExistingNodeValue +                                # ACV OJO FIX replacing line 
                         [node.firstChild.nodeValue]
                       }
                    )                
            else:
                id=node.getAttribute('id')
                uid=node.getAttribute('uid')
                
                typename_=[tagname[0]]
                for c in tagname[1:]:
                    if c.isupper():
                        typename_.append(' '+c)
                    else:
                        typename_.append(c)
                        
                typename=str(''.join(typename_))
                
                #we are in overwrite mode, so we delete the existing subobjects
                if hasattr(self.aq_base,id):
                    self.manage_delObjects([id])
                    
                self.invokeFactory(typename,id)
                obj=getattr(self,id)

                if uid:
                    doc.objectsByUid[uid]=obj

                if hasattr(obj.aq_base,'importDOM'):
                    obj.importDOM(doc, node, root=0)
                    
                    
                #restore the references
                if root:
                    for r in doc.references:
                        target=doc.objectsByUid[r[2]]
                        ssource=r[0]
                        field=r[1]
                        field.set(ssource,target)
                        
            # check for subtransaction
            if counter % IMPORT_TRANSACTION_STEPPING == 0:
                transaction.commit(1)
            
    def exportXML(self):
        doc=minidom.Document()
        
        self.exportDOM(doc,doc)
        return doc.toprettyxml()
    
    def exportDOM(self,doc,dom):
        classNode=doc.createElement(self.portal_type.replace(' ',''))
        classNode.setAttribute('id',self.id)
        classNode.setAttribute('title',self.Title())
        classNode.setAttribute('uid',self.UID())

        fields=[f for f in self.Schema().fields() if f.getName() \
             not in ('id','title','xml') and not f.isMetadata]

        for field in fields:
            if field.type=='lines':
                values=field.get(self)
                for value in values:
                    node=doc.createElement(field.getName()[:-1])
                    textnode=doc.createTextNode(value)
                    node.appendChild(textnode)
                    classNode.appendChild(node)
            elif field.type=='reference':
                node=doc.createElement(field.getName())
                node.setAttribute('uidref',field.getRaw(self))
                classNode.appendChild(node)
            else:
                node=doc.createElement(field.getName())
                textnode=doc.createTextNode(str(field.get(self)))
                node.appendChild(textnode)
                classNode.appendChild(node)
            
        for o in self.objectValues():
            o.exportDOM(doc,classNode)
            
        dom.appendChild(classNode)

class ReferenceLayerManager(Reference):
    """Forwards hooks to ReferenceLayers, see IReferenceLayerProvider."""
    
    def addHook(self, tool, sourceObject=None, targetObject=None):
        self._forEachLayerDo('addHook', self)

    def delHook(self, tool, sourceObject=None, targetObject=None):
        self._forEachLayerDo('delHook', self)

    def beforeTargetDeleteInformSource(self):
        self._forEachLayerDo('beforeTargetDeleteInformSource', self)

    def beforeSourceDeleteInformTarget(self):
        self._forEachLayerDo('beforeSourceDeleteInformTarget', self)

    def getRuleset(self):
        if not getattr(self, '_v_ruleset', None):
            library = getToolByName(self, RELATIONS_LIBRARY)
            self._v_ruleset = library.getRuleset(self.relationship)
        return self._v_ruleset

    def _forEachLayerDo(self, methodName, *args, **kwargs):
        layers = self._getReferenceLayers()
        for l in layers:
            method = getattr(l, methodName, None)
            if callable(method):
                method(*args, **kwargs)

    def _getReferenceLayers(self):
        rs = self.getRuleset()
        providers = rs.getComponents(interfaces.IReferenceLayerProvider)
        return [p.provideReferenceLayer(self) for p in providers]

InitializeClass(ReferenceLayerManager)

class RLMWithBrains(ReferenceLayerManager, brain.ReferenceWithBrains):
    pass

class RuleBase(XMLImportExport):
    __implements__ = interfaces.IRule,

    global_allow = 0 # convenience
    def getRuleset(self): return aq_parent(aq_inner(self))

class DefaultPrimaryImplicator(RuleBase):
    __implements__ = interfaces.IPrimaryImplicator,

    referenceClass = RLMWithBrains

    def __init__(self, ruleset):
        self.ruleset = ruleset
    
    def connect(self, source, target, metadata=None):
        if metadata is None:
            metadata = {}
        
        ref_catalog = getToolByName(self.ruleset, REFERENCE_CATALOG)

        args = (source.UID, target.UID, self.ruleset.getId(),
                self.referenceClass)
        kwargs = metadata
        if config.ALLOW_MULTIPLE_REFS_PER_TRIPLE:
            kwargs['updateReferences'] = False
        
        return ref_catalog.addReference(*args, **kwargs)

    def disconnect(self, reference):
        ref_catalog = getToolByName(self.ruleset, REFERENCE_CATALOG)
        ref_catalog._deleteReference(reference)
        
class Ruleset(utils.AllowedTypesByIface, OrderedBaseFolder, XMLImportExport):
    """See IRuleset."""
    __implements__ = (interfaces.IRuleset,) +\
                     OrderedBaseFolder.__implements__

    schema = schema.RulesetSchema
    portal_type = archetype_name = 'Ruleset'
    global_allow = 0
    content_icon = 'ruleset_icon.gif'
    allowed_interfaces = interfaces.IRule, #used by AllowedTypesByIface

##     def initializeArchetype(self, **kwargs):
##         OrderedBaseFolder.initializeArchetype(self, **kwargs)

    def getComponents(self, interface):
        """Return a list of objects in self that implement the given
        interface."""
        return [obj for obj in self.objectValues()
                if interface.isImplementedBy(obj)]

    def implyOnConnect(self, source, target, chain, metadata=None):
        primaryImplicator = self._getPrimaryImplicator()
        reference = primaryImplicator.connect(source, target, metadata)
        chain.added.append(reference)
        self.addReference(reference, RELATIONSHIP_RULESETTOREF)
        self._forEachDo(interfaces.IImplicator, 'implyOnConnect',
                        reference, chain)

    def implyOnDisconnect(self, reference, chain):
        primaryImplicator = self._getPrimaryImplicator()
        primaryImplicator.disconnect(reference)
        chain.deleted.append(reference)
##         reference = ImplicitAcquisitionWrapper(reference, self)
        self._forEachDo(interfaces.IImplicator, 'implyOnDisconnect',
                        reference, chain)

    def validateConnected(self, reference, chain):
        self._forEachDo(interfaces.IValidator, 'validateConnected',
                        reference, chain)

    def validateDisconnected(self, reference, chain):
        self._forEachDo(interfaces.IValidator, 'validateDisconnected',
                        reference, chain)

    def finalizeOnConnect(self, reference, chain):
        self._forEachDo(interfaces.IFinalizer, 'finalizeOnConnect',
                        reference, chain)

    def finalizeOnDisconnect(self, reference, chain):
        self._forEachDo(interfaces.IFinalizer, 'finalizeOnDisconnect',
                        reference, chain)

    def makeVocabulary(self, source, targets=None):
        iface = interfaces.IVocabularyProvider
        for obj in self.getComponents(iface):
            targets = obj.makeVocabulary(source, targets)
        return targets

    def listActionsFor(self, reference):
        actions = []
        iface = interfaces.IReferenceActionProvider
        for obj in self.getComponents(iface):
            actions = actions + obj.listActionsFor(reference)

        seen_urls = []
        value = []
        for action in actions:
            if action['url'] not in seen_urls:
                seen_urls.append(action['url'])
                value.append(action)
        return value

    def _forEachDo(self, interface, methodname, *args):
        """For each contained object that is implements interface, call method
        with *args."""
        for obj in self.getComponents(interface):
            meth = getattr(obj.aq_explicit, methodname)
            meth(*args)

    def _getPrimaryImplicator(self):
        impls = self.getComponents(interfaces.IPrimaryImplicator)
        if impls:
            return impls[0]
        else:
            return DefaultPrimaryImplicator(self)

    def _afterRename(self, context):
        """We have been renamed -> set relationship attributes on refs."""
        ref_ctl = getToolByName(context, REFERENCE_CATALOG)
        refs = self.__of__(context).getRefs(RELATIONSHIP_RULESETTOREF)
        for ref in refs:
            ref.relationship = self.getId()
            url = getRelURL(aq_parent(aq_inner(ref)), ref.getPhysicalPath())
            ref_ctl.catalog_object(ref, url, idxs=['relationship'])

registerType(Ruleset)


class RulesetAwareContainer:
    """Mix-in that's responsible for adding a reference from the
    library to new rulesets and informing rulesets when they get
    renamed.

    XXX: Depends on subclasses to also inherit from
    AllowedTypesByIface and OrderedBaseFolder."""

    # Add reference to Rulesets on invokeFactory
    def invokeFactory(self, type_name, id, RESPONSE = None, *args, **kwargs):
        library = getToolByName(self, RELATIONS_LIBRARY)

        super_invokeFactory = utils.AllowedTypesByIface.invokeFactory
        v = super_invokeFactory(self, type_name, id, RESPONSE, *args, **kwargs)

        obj = getattr(self, v)
        if interfaces.IRuleset.isImplementedBy(obj):

            library.addReference(obj, RELATIONSHIP_LIBRARY)
        return v

    # This hack allows us to inform the ruleset that it has been renamed.
    def _setObject(self, id, obj, roles=None, user=None, set_owner=1):
        library = getToolByName(self, RELATIONS_LIBRARY)
        
        super_setObject = OrderedBaseFolder._setObject
        super_setObject(self, id, obj, roles, user, set_owner)

        if interfaces.IRuleset.isImplementedBy(obj):
            ruleset = obj
            ref_ctl = getToolByName(self, REFERENCE_CATALOG)
            brains = ref_ctl(sourceUID=library.UID(),
                             targetUID=ruleset.UID(),
                             relationship=RELATIONSHIP_LIBRARY)
            if len(brains) != 0: # We assume a rename
                ruleset._afterRename(self)
            else: # Assume a copy: we need to register the ruleset
                library.addReference(ruleset, RELATIONSHIP_LIBRARY)


class Library(RulesetAwareContainer, utils.AllowedTypesByIface,
              ActionProviderBase, OrderedBaseFolder, XMLImportExport):
    """Registry for IRulesets. See ILibrary."""
    __implements__ = ((interfaces.ILibrary,) +
                      (ActionProviderBase.__implements__,
                       OrderedBaseFolder.__implements__, ))
                      

    schema = schema.BaseSchemaWithInvisibleId + XMLImportExport.schema
    portal_type = archetype_name = 'Relations Library'
    global_allow = 0
    content_icon = 'library_icon.gif'
    allowed_interfaces = interfaces.IRuleset, interfaces.IRulesetCollection

    def initializeArchetype(self, **kwargs):
        # add action
        self.addAction(
            id='relations',
            name='Relations',
            action='string:$object_url/relations_form',
            condition='python:object.relations_adddelete_vocab(test_only=1)',
            permission=('Modify portal content',),
            category='object_tabs',
            visible=1)
        
        OrderedBaseFolder.initializeArchetype(self, **kwargs)
        
    ##
    # see ILibrary
    def registerRuleset(self, ruleset):
        self.addReference(ruleset, RELATIONSHIP_LIBRARY)

    def getRuleset(self, id):
        ref_ctl = getToolByName(self, REFERENCE_CATALOG)
        uid_ctl = getToolByName(self, UID_CATALOG)

        # XXX: targetId index of reference_catalog is not being updated
        refs = ref_ctl(sourceUID=self.UID(),
                       relationship=RELATIONSHIP_LIBRARY)
        rulesets = uid_ctl(UID=[r.targetUID for r in refs], id=id)
        if rulesets:
            return rulesets[0].getObject()
        else:
            raise ValueError, "No ruleset with id %r" % id

    def getRulesets(self):
        # We want to have a reasonable order of rulesets.  First come the
        # rulesets in self, then rulesets of collections.
        v = self.objectValues('Ruleset')
        for collection in self.objectValues('Ruleset Collection'):
            v = v + collection.getRulesets()

        # look for objs that are not contained in self
        for obj in self.getRefs(RELATIONSHIP_LIBRARY):
            if obj not in v:
                v.append(obj)
        return v

    def getFolder(self):
        return self

registerType(Library)


class RulesetCollection(RulesetAwareContainer, utils.AllowedTypesByIface,
                        OrderedBaseFolder, XMLImportExport):
    """A container for IRulesets that lives inside the library."""
    __implements__ = (interfaces.IRulesetCollection,) + \
                     OrderedBaseFolder.__implements__

    schema = schema.BaseSchemaWithInvisibleId + XMLImportExport.schema
    portal_type = archetype_name = 'Ruleset Collection'
    global_allow = 0
    allowed_interfaces = interfaces.IRuleset, interfaces.IRulesetCollection

    def getRulesets(self):
        v = self.objectValues('Ruleset')
        for collection in self.objectValues('Ruleset Collection'):
            v = v + collection.getRulesets()
        return v

registerType(RulesetCollection)
